### ISSUE TITLE
 .htpasswd authentication not working when using allow_unauthenticated_reads: true

### ISSUE THREAD

**Issue by Snapstromegon** [buchgr/bazel-remote#468]
We're running the docker image on version 2.1.2 in our environment and detected a problem with the *allow_unauthenticated_reads: true* feature.

**Used version of the image:**
```sh
$ docker images
buchgr/bazel-remote-cache   latest    c2afbbddcd67
```

**Docker run command to start the image:**
```$sh
docker run -p 8080:8080 -v ~/.config/bazel-remote/:/config buchgr/bazel-remote-cache --config_file /config/cache.yaml
```

When we now use basic auth (with a .htpasswd file provided in the config) to connect to the cache from bazel, we get 401 errors for "PUT" requests.
This does not happen when *allow_unauthenticated_reads* is disabled and everything else is the same.

**Comment by mostynb** [comment]
Hi, could you share your config file and bazel-remote's startup log?

**Comment by Snapstromegon** [comment]
Here are the logs and the config

Startup log:

```sh
[ec2-user@ip-<redacted> ~]$ docker run -p 8080:8080 -v ~/.config/bazel-remote/:/config buchgr/bazel-remote-cache --config_file /config/cache.yaml
2021/09/09 10:09:43 bazel-remote built with go1.16.2 from git commit ee83b71253a191f8fbf5a95f06758c0986dc4bfe.
Using S3 backend.
2021/09/09 10:09:43 Initial RLIMIT_NOFILE cur: 32768 max: 65536
2021/09/09 10:09:43 Setting RLIMIT_NOFILE cur: 65536 max: 65536
2021/09/09 10:09:43 Loading existing files in /home/ec2-user/cache.
2021/09/09 10:09:43 Sorting cache files by atime.
2021/09/09 10:09:43 Building LRU index.
2021/09/09 10:09:43 Finished loading disk cache files.
2021/09/09 10:09:43 Loaded 0 existing disk cache items.
2021/09/09 10:09:43 Authentication: basic
2021/09/09 10:09:43 Access mode: authentication required for writes, unauthenticated reads allowed
2021/09/09 10:09:43 Mangling non-empty instance names with AC keys: disabled
2021/09/09 10:09:43 Starting HTTP server on address 0.0.0.0:8080
2021/09/09 10:09:43 HTTP AC validation: enabled
[...]
2021/09/09 10:10:14 S3 DOWNLOAD <redacted> <redacted>/ac/31/311ee57031836b5717331ceaebd7990254f32496cb1b92f9865c764bc1ec4fa5 NOT FOUND
2021/09/09 10:10:14  GET 404     <redacted> /ac/311ee57031836b5717331ceaebd7990254f32496cb1b92f9865c764bc1ec4fa5
2021/09/09 10:10:18  PUT 401     <redacted> /cas/4042d72b9e413c66e6ae9976e9db6ca12692bb4d2c0e744e611e1a394d32d102
2021/09/09 10:10:18  PUT 401     <redacted> /cas/2d08f6b55a8067a82b4b9710c031aefe3aea93407a4973cc975368f009d2edbd
2021/09/09 10:10:18  PUT 401     <redacted> /cas/b021bd2130ba7d9f33ba1db2108b3112f6156f1a84956764349427974f220914
2021/09/09 10:10:18  PUT 401     <redacted> /cas/3cccb4056a13a6d813171efc4088a8a729e4cb0735889d81a2f48ae1a87eaae9
2021/09/09 10:10:18  PUT 401     <redacted> /cas/76886956a9113a5112ebf1cf1054935516c0b411fc6f1102deb58c88842f6112
[...]
```

Short Version of config (all "#" lines removed):

```conf
dir: /home/ec2-user/cache
max_size: 25
host: 0.0.0.0
port: 8080
http_read_timeout: 15s
http_write_timeout: 20s
htpasswd_file: /config/.htpasswd
allow_unauthenticated_reads: true
key_format: 2
s3_proxy:
  endpoint: s3.eu-central-1.amazonaws.com
  bucket: <redacted>
  prefix: <redacted>
  access_key_id: <redacted>
  secret_access_key: <redacted>
```

<details>
<summary>Long Version of config</summary>

```conf
# These two are the only required options:
dir: /home/ec2-user/cache
max_size: 25

# The form to store CAS blobs in ("zstd" or "uncompressed"):
#storage_mode: zstd

host: 0.0.0.0
# The port to use for HTTP/HTTPS:
port: 8080
# The port to use for gRPC:
#grpc_port: 9092

# If profile_port is specified, then serve /debug/pprof/* URLs here:
#profile_host: 127.0.0.1
#profile_port: 7070

# HTTP read/write timeouts. Note that these do not apply to the proxy
# backends or the profiling endpoint. Reasonable values might be twice
# the length of time that you expect a client to read/write the largest
# likely blob. Units can be one of: "s", "m", "h".
http_read_timeout: 15s
http_write_timeout: 20s

# Specify a certificate if you want to use HTTPS:
#tls_cert_file: path/to/tls.cert
#tls_key_file:  path/to/tls.key
# If you want to use mutual TLS with client certificates:
#tls_ca_file: path/to/ca/cert.pem

# Alternatively, you can use simple authentication:
htpasswd_file: /config/.htpasswd

# If tls_ca_file or htpasswd_file are specified, you can choose
# whether or not to allow unauthenticated read access:
allow_unauthenticated_reads: true

#To avoid per-prefix rate limiting with Amazon S3 enable:
key_format: 2
# If specified, bazel-remote should exit after being idle
# for this long. Time units can be one of: "s", "m", "h".
#idle_timeout: 45s

# If set to true, do not validate that ActionCache
# items are valid ActionResult protobuf messages.
#disable_http_ac_validation: false

# If set to true, do not check that CAS items referred
# to by ActionResult messages are in the cache.
#disable_grpc_ac_deps_check: false

# If set to true, enable metrics for each HTTP/gRPC endpoint.
#enable_endpoint_metrics: false

# Specify a custom list of histogram buckets for endpoint request duration metrics
#endpoint_metrics_duration_buckets: [.5, 1, 2.5, 5, 10, 20, 40, 80, 160, 320]

# At most one of the proxy backends can be selected:
#
# If this is 0, proxy backends won't upload blobs.
#num_uploaders: 100
# The maximum number of proxy uploads to queue, before dropping uploads.
#max_queued_uploads: 1000000
# The largest blob size that will be accepted, for example 10MB:
#max_blob_size: 10485760
#
#gcs_proxy:
#  bucket: gcs-bucket
#  use_default_credentials: false
#  json_credentials_file: path/to/creds.json
#
s3_proxy:
  endpoint: s3.eu-central-1.amazonaws.com
  bucket: <redacted>
  prefix: <redacted>
  access_key_id: <redacted>
  secret_access_key: <redacted>
#
# Provide either access_key_id/secret_access_key, or iam_role_endpoint/region.
# iam_role_endpoint can also be left empty, and figured out automatically.
#  iam_role_endpoint: http://169.254.169.254
#  region: us-east-1
#
#http_proxy:
#  url: https://remote-cache.com:8080/cache

# If set to a valid port number, then serve /debug/pprof/* URLs here:
#profile_port: 7070
# IP address to use, if profiling is enabled:
#profile_host: 127.0.0.1

# If true, enable experimental remote asset API support:
#experimental_remote_asset_api: true

# If supplied, controls the verbosity of the access logger ("none" or "all"):
#access_log_level: none
```
</details>

**Comment by mostynb** [comment]
I have a potential fix in #469 - but I still need to add some regression tests before landing.

Could you try this in the meantime, and report if it works for you?
https://hub.docker.com/layers/mostynb/bazel-remote-cache/fix_basic_http_auth_writes/images/sha256-3d6aea11b1adadbbd0f6cca7083541d8e4654dc8079355d047dbe79b24d2a389?context=repo

**Comment by minmichail** [comment]
> I have a potential fix in #469 - but I still need to add some regression tests before landing.
> 
> Could you try this in the meantime, and report if it works for you?
> https://hub.docker.com/layers/mostynb/bazel-remote-cache/fix_basic_http_auth_writes/images/sha256-3d6aea11b1adadbbd0f6cca7083541d8e4654dc8079355d047dbe79b24d2a389?context=repo

hi,
I tested it right now: 
"allow_unauthenticated_reads = true" => PUT requests are not anymore shown in logs, also nothing is available on machine/s3
"allow_unauthenticated_reads=false" => PUT requests are shown in the logs and everything is working properly

**Comment by Snapstromegon** [comment]
@mostynb if I read @minmichail 's comment correctly, the cache isn't working at all, since the request seems to get lost. I don't think this resolves the issue - at least if I read @minmichail 's comment correctly.

**Comment by mostynb** [comment]
Sorry- I created some tests, then reworked the PR with the tests passing now. There's a new image with these updates that you can try here (I was waiting for that to finish before leaving an update here):

https://hub.docker.com/layers/buchgr/bazel-remote-cache/latest/images/sha256-b5d8a7d5ae0f647444d6ea41017c57b7b41b5fbdebc342aef365924ae57b96a8?context=repo

(Reopening the issue until the fix has been confirmed...)

**Comment by minmichail** [comment]
issue seems to be fixed.

Thanks for your quick support  @mostynb

**Comment by mostynb** [comment]
Thanks for checking- released v2.1.3 (and pushed it to docker hub).
